internal/querynodev2/delegator/delegator.go

```diff
@@ -203,6 +207,140 @@ func (sd *shardDelegator) modifyQueryRequest(req *querypb.QueryRequest, scope qu
+		log.Debug("skip optimize based on clustering info")
+		return req, sealeds
+	}
+
+	// basic optimize rule
+	topK := req.GetReq().GetTopk()
+	filterRatio := req.GetReq().GetClusteringOptions().GetFilterRate()
+	dim := req.GetReq().GetDim()
+	metricType := req.GetReq().GetMetricType()
+	var phg commonpb.PlaceholderGroup
+	err := proto.Unmarshal(req.GetReq().GetPlaceholderGroup(), &phg)
+	if err != nil {
+		log.Warn("fail to parse SearchRequest PlaceholderGroup", zap.Error(err))
+		return req, sealeds
+	}
+	log.Debug("optimizeSearchBasedOnClustering",
+		zap.String("metricType", metricType),
+		zap.Int32("dim", dim),
+		zap.Int("length", len(phg.GetPlaceholders())),
+		zap.Any("phg", phg))
+
+	vectorsBytes := phg.GetPlaceholders()[0].GetValues()
+	searchVectors := make([][]float32, len(vectorsBytes))
+	for i, vectorBytes := range vectorsBytes {
+		searchVectors[i] = clustering.DeserializeFloatVector(vectorBytes)
+	}
+
+	var segments = make([]SegmentEntry, 0)
+	for _, sealed := range sealeds {
+		segments = append(segments, sealed.Segments...)
+	}
+
+	type segmentDistanceStruct struct {
+		segment  SegmentEntry
+		distance float32
+	}
+
+	var vectorSegmentDistances = make([][]segmentDistanceStruct, 0)
+	for _, searchVector := range searchVectors {
+		var vectorSegmentDistance = make([]segmentDistanceStruct, 0)
+		for _, segment := range segments {
+
+			if segment.ClusteringInfos != nil {
+				for _, clusteringInfo := range segment.ClusteringInfos {
+					distance, err := distance.CalcFloatDistance(int64(dim), searchVector, clusteringInfo.Centroid, metricType)
+					if err != nil {
+						log.Error("Fail to calculate distance between clustering center and search vector", zap.Error(err))
+					}
+					log.Debug("distance between searchVector and cluster center",
+						zap.Int64("segmentID", segment.SegmentID),
+						zap.Float32s("distance", distance),
+						zap.Float32s("searchVector", searchVector),
+						zap.Float32s("clusterCentroid", clusteringInfo.Centroid))
+					if len(distance) > 0 {
+						vectorSegmentDistance = append(vectorSegmentDistance, segmentDistanceStruct{segment: segment, distance: distance[0]})
+					} else {
+						// if no legal distance is calculated, regard this segment as a normal one.
+						vectorSegmentDistance = append(vectorSegmentDistance, segmentDistanceStruct{segment: segment, distance: float32(0.0)})
+					}
+				}
+			} else {
+				vectorSegmentDistance = append(vectorSegmentDistance, segmentDistanceStruct{segment: segment, distance: float32(0.0)})
+			}
+		}
+		vectorSegmentDistances = append(vectorSegmentDistances, vectorSegmentDistance)
+	}
+	log.Debug("vector segment distances", zap.Any("vectorSegmentDistances", vectorSegmentDistances))
+
+	optimizedSegs := make(map[UniqueID]SegmentEntry, 0)
+	for _, vectorSegmentDistance := range vectorSegmentDistances {
+		sort.SliceStable(vectorSegmentDistance, func(i, j int) bool {
+			return vectorSegmentDistance[i].distance < vectorSegmentDistance[j].distance
```

Comment:
for IP or cosine metric, larger result means more similar, we should sort reverse